runtime/commonMain/src/kotlinx/serialization/SerialDescriptor.kt

```diff
@@ -0,0 +1,188 @@
+     * the serializer itself. // TODO #363
+     */
+    public val serialName: String
+
+    /**
+     * The kind of the serialized form that determines **the shape** of the serialized data.
+     * Formats use serial kind to add and parse serializer-agnostic metadata to the result.
+     *
+     * For example, JSON format wraps [classes][StructureKind.CLASS] and [StructureKind.MAP] into
+     * brackets, while ProtoBuf just serialize these types in a separate ways.
+     *
+     * Kind should be consistent with the implementation, for example, if it is a [primitive][PrimitiveKind],
+     * then its elements count should be zero and vice versa.
+     */
+    public val kind: SerialKind
+
+    /**
+     * Whether the descriptor describes nullable element.
+     * Returns `true` if associated serializer can serialize/deserialize nullable elements of the described type.
+     */
+    public val isNullable: Boolean get() = false
+
+    /**
+     * The number of elements this descriptor describes, besides from the element itself.
+     * [elementsCount] describes the number of **semantic** elements, not the number
+     * of actual fields/properties in the serialized form, even though they frequently match.
+     *
+     * For example, for the following class
+     * `class Complex(val real: Long, val imaginary: Long)` the corresponding descriptor
+     * and the serialized form both have two elements, while for `class IntList : ArrayList<Int>()`
+     * the corresponding descriptor has a single element (`IntDescriptor`, the type of list element),
+     * but from zero up to `Int.MAX_VALUE` elements in the serialized form.
+     */
+    public val elementsCount: Int
+
+    /**
+     * Returns a serial annotations of the associated class.
+     * Serial annotations can be used to specify an additional
+     * metadata that may be used during serialization, for example a [serial name][SerialName].
+     *
+     * TODO `serialAnnotations` or get rid of it completely?
+     * TODO default shouldn't be here either?
+     */
+    public val annotations: List<Annotation> get() = emptyList()
+
+    /**
+     * Returns a _positional_ name of the child at the given [index].
+     * Positional name usually represents a corresponding property name in the class, associated with
+     * the current descriptor.
+     *
+     * @throws IndexOutOfBoundsException for an illegal [index] values.
+     * @throws IllegalStateException if the current descriptor does not support children elements (e.g. is a primitive)
+     */
+    public fun getElementName(index: Int): String
+
+    /**
+     * Returns an index in the children list of the given element by its name or [CompositeDecoder.UNKNOWN_NAME]
+     * if there is no such element.
+     * The resulting index, if it is not [CompositeDecoder.UNKNOWN_NAME], is guaranteed to be usable with
+     * [getElementAnnotations], [getElementName] and [isElementOptional]
+     *
+     * TODO MapLike does not comply this contract
```

Comment:
Both `MapLike` and `ListLike` does not comply in `getElementAnnotations` part, but comply in `getElementDescriptor`, which you didn't mention. I think `getElementDescriptor` method is worth some clarifying.